File: LeetCode/April Challenge/16th Day/Day_16.cpp
class Solution {
public:
bool checkValidString(string s) {
      stack<int> open,ast;
      for(int i=0;i<s.length();i++)
      {
          if(s[i] == ')')
          {
              if(!open.empty()) open.pop();
              else if(!ast.empty()) ast.pop();
              else return false;
          }
          else if(s[i] == '(')   open.push(i);
          else ast.push(i);
      }
      while(!open.empty() && !ast.empty())
      {
         if(open.top() > ast.top()) return false;
         open.pop();
          ast.pop();
      }
        return open.empty();
    }
};

File: LeetCode/April Challenge/16th Day/Day_16.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        left_paren_index = []
        star_index = []
        for i, val in enumerate(s):
            if val == "(":
                left_paren_index.append(i)
            elif val == ")":
                if left_paren_index:
                    left_paren_index.pop()
                elif star_index:
                    star_index.pop()
                else:
                    return False
            else:
                star_index.append(i)
        while left_paren_index:
            if len(star_index) == 0 or star_index.pop() < left_paren_index.pop():
                return False
        return True
